### PR TITLE
Enhancements to password management

### PR DESCRIPTION
**Pull request by taohe1012**
##### SUMMARY
Requirements
Once customer uses the Redfish User Account resource to change the root password, then they will have to manually edit the state files to change the old root password to new password. If they fail to do this, then a subsequent Terraform Plan/Apply will lead to errors as the old password, which is stored in the state files, is used for the Read Refresh operations. In the field, root password is frequently rotated so we need to find a way to help customers avoid this issue after the root password change.

Solution
1.  Before every resource/datasource is given server details which are stored in the their config/state  ---> now also cache entire server details map into provider ("redfish_servers").

2.  Before the connection is created with server details from resource/datasource directly ---> now the connection will use an identifier("redfish_alias") for each iDRAC to get server details from provider firstly; if the "redfish_alias" is null, then use server details from resource/datasource to create connection.

**Note: Changes for document, example files and AT cases will be in next PR.**
##### ISSUE TYPE

- Bugfix Pull Request

##### ADDITIONAL INFORMATION

```paste below

```

#### Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note

```

#### Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

#### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

### CODE REVIEW

**Comment by terraform-providers-svc** [comment]
Can one of the admins verify this patch?

File: redfish/provider/common.go [hunk 1]
```diff
@@ -183,25 +194,25 @@ func NewConfig(pconfig *redfishProvider, rserver *[]models.RedfishServer) (*gofi
 		return nil, fmt.Errorf("no provider block was found")
 	}
 
-	// first redfish server block
-	if len(*rserver) == 0 {
```
Comment (rounak-adhikary) [inline redfish/provider/common.go:187]: This check is required, rt? We do still want to throw an error if user doesn't provide the `redfish_server` block.

File: redfish/provider/common.go [hunk 2]
```diff
@@ -76,6 +77,11 @@ func RedfishServerSchema() map[string]resourceSchema.Attribute {
 			Optional:    true,
 			Description: "This field indicates whether the SSL/TLS certificate must be verified or not",
 		},
+		"redfish_alias": resourceSchema.StringAttribute{
+			MarkdownDescription: redfishAliasMD,
+			Description:         redfishAliasMD,
+			Optional:            true,
```
Comment (rounak-adhikary) [inline redfish/provider/common.go:83]: Should we put validators to check that redfish_alias conflicts with all other fields?

File: redfish/provider/common.go (hunk 1 above)
Comment (taohe1012) [reply redfish/provider/common.go:187]: hi @rounak-adhikary there are two same check. I removed one of them. Please help confirm the double check is necessary.

File: redfish/provider/common.go (hunk 2 above)
Comment (taohe1012) [reply redfish/provider/common.go:83]: @rounak-adhikary , user/password are optional, it's ok to use conflicts , but the endpoint required, we cannot use conflicts for it. So thinking to keep user/password and endpoint  consistency, I not use the conflicts for them here.

File: redfish/provider/common.go (hunk 1 above)
Comment (rounak-adhikary) [reply redfish/provider/common.go:187]: Oops. The missed the extra check above.
U are rt. The double check isn't necessary. Thanks for removing it.

File: redfish/provider/common.go (hunk 2 above)
Comment (rounak-adhikary) [reply redfish/provider/common.go:83]: Ah yes, we cannot conflict with a required field.
But we also cannot have users provide both `redfish_alias` and `endpoint` at the same time as that would be an invalid input.
So can we demote `endpoint` to an optional field, and put a validator saying either the `endpoint` or `redfish` alias is required?

Comment (taohe1012) [reply redfish/provider/common.go:83]: sure @rounak-adhikary , we can do this. Will demote endpoint to optional and add validation that  one of them is required. Thanks.

Comment (taohe1012) [reply redfish/provider/common.go:93]: hi @rounak-adhikary , the improvement is added. Please help review.
And i run the AT for user resource, all existing at cases are passed.
image

**Review by rounak-adhikary — APPROVED** [review]
LGTM

**Review by Krishnan-Priyanshu — APPROVED** [review]

**Pull request merged**